File: scripts/validate.js
#!/usr/bin/env node

/**
 * validate.js - Repository Structure Validator
 * 
 * Validates that all tool directories follow the required structure
 * and that files are properly formatted.
 */

const fs = require('fs');
const path = require('path');

// Configuration
const REQUIRED_STRUCTURE = {
  toolDir: {
    'README.md': true, // Required in each tool directory
    'Prompt.txt': false, // At least one prompt file required (can be versioned)
    'Tools.json': false, // Optional
  }
};

const EXCLUDED_DIRS = ['.git', '.github', 'site', 'assets', 'node_modules', 'scripts', 'metadata'];
const ROOT_DIR = path.join(__dirname, '..');

// Colors for terminal output
const colors = {
  reset: '\x1b[0m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
};

class Validator {
  constructor() {
    this.errors = [];
    this.warnings = [];
    this.passed = 0;
  }

  log(message, color = 'reset') {
    console.log(`${colors[color]}${message}${colors.reset}`);
  }

  error(message) {
    this.errors.push(message);
    this.log(`❌ ERROR: ${message}`, 'red');
  }

  warn(message) {
    this.warnings.push(message);
    this.log(`⚠️  WARNING: ${message}`, 'yellow');
  }

  success(message) {
    this.passed++;
    this.log(`✅ ${message}`, 'green');
  }

  info(message) {
    this.log(`ℹ️  ${message}`, 'blue');
  }

  // Get all tool directories
  getToolDirectories() {
    const dirs = fs.readdirSync(ROOT_DIR, { withFileTypes: true })
      .filter(dirent => dirent.isDirectory())
      .map(dirent => dirent.name)
      .filter(name => !EXCLUDED_DIRS.includes(name) && !name.startsWith('.'));
    
    return dirs;
  }

  // Check if directory has at least one prompt file
  hasPromptFile(dirPath) {
    const files = fs.readdirSync(dirPath);
    return files.some(file => 
      file.toLowerCase().includes('prompt') && 
      (file.endsWith('.txt') || file.endsWith('.md'))
    );
  }

  // Validate tool directory structure
  validateToolDirectory(toolName) {
    const toolPath = path.join(ROOT_DIR, toolName);
    const files = fs.readdirSync(toolPath);

    let valid = true;

    // Check for README.md
    if (!files.includes('README.md')) {
      this.warn(`${toolName}: Missing README.md`);
      valid = false;
    }

    // Check for at least one prompt file
    if (!this.hasPromptFile(toolPath)) {
      this.error(`${toolName}: No prompt file found`);
      valid = false;
    }

    if (valid) {
      this.success(`${toolName}: Structure valid`);
    }

    return valid;
  }

  // Validate file content
  validatePromptFile(toolName, fileName) {
    const filePath = path.join(ROOT_DIR, toolName, fileName);
    const content = fs.readFileSync(filePath, 'utf-8');

    // Check if file is not empty
    if (content.trim().length === 0) {
      this.error(`${toolName}/${fileName}: File is empty`);
      return false;
    }

    // Check for common issues
    if (content.includes('[REDACTED]') && !content.includes('redaction')) {
      this.warn(`${toolName}/${fileName}: Contains [REDACTED] markers`);
    }

    // Check file size (warn if suspiciously small)
    if (content.length < 500) {
      this.warn(`${toolName}/${fileName}: Suspiciously small (${content.length} chars)`);
    }

    return true;
  }

  // Validate JSON files
  validateJsonFile(toolName, fileName) {
    const filePath = path.join(ROOT_DIR, toolName, fileName);
    
    try {
      const content = fs.readFileSync(filePath, 'utf-8');
      JSON.parse(content);
      this.success(`${toolName}/${fileName}: Valid JSON`);
      return true;
    } catch (error) {
      this.error(`${toolName}/${fileName}: Invalid JSON - ${error.message}`);
      return false;
    }
  }

  // Validate README.md in tool directory
  validateToolReadme(toolName) {
    const readmePath = path.join(ROOT_DIR, toolName, 'README.md');
    
    if (!fs.existsSync(readmePath)) {
      return false; // Already warned in validateToolDirectory
    }

    const content = fs.readFileSync(readmePath, 'utf-8');

    // Check for required sections (basic)
    const requiredSections = ['Overview', 'Files'];
    const missingSections = requiredSections.filter(section => 
      !content.toLowerCase().includes(section.toLowerCase())
    );

    if (missingSections.length > 0) {
      this.warn(`${toolName}/README.md: Missing sections: ${missingSections.join(', ')}`);
    }

    return true;
  }

  // Check for duplicate tool names (case-insensitive)
  checkDuplicates() {
    const dirs = this.getToolDirectories();
    const lowerCaseNames = dirs.map(d => d.toLowerCase());
    const duplicates = lowerCaseNames.filter((name, index) => 
      lowerCaseNames.indexOf(name) !== index
    );

    if (duplicates.length > 0) {
      this.error(`Duplicate tool directories found: ${duplicates.join(', ')}`);
    }
  }

  // Main validation
  run() {
    this.log('\n🔍 Starting Repository Validation...\n', 'blue');

    // Get all tool directories
    const toolDirs = this.getToolDirectories();
    this.info(`Found ${toolDirs.length} tool directories\n`);

    // Validate each tool directory
    toolDirs.forEach(toolName => {
      this.validateToolDirectory(toolName);
      
      // Validate specific files if they exist
      const toolPath = path.join(ROOT_DIR, toolName);
      const files = fs.readdirSync(toolPath);

      // Validate prompt files
      files.filter(f => f.toLowerCase().includes('prompt') && f.endsWith('.txt'))
        .forEach(file => this.validatePromptFile(toolName, file));

      // Validate JSON files
      files.filter(f => f.endsWith('.json'))
        .forEach(file => this.validateJsonFile(toolName, file));

      // Validate README
      this.validateToolReadme(toolName);
    });

    // Check for duplicates
    this.checkDuplicates();

    // Print summary
    this.printSummary();

    // Exit with appropriate code
    process.exit(this.errors.length > 0 ? 1 : 0);
  }

  printSummary() {
    this.log('\n' + '='.repeat(50), 'blue');
    this.log('📊 Validation Summary', 'blue');
    this.log('='.repeat(50) + '\n', 'blue');

    this.log(`✅ Passed: ${this.passed}`, 'green');
    this.log(`⚠️  Warnings: ${this.warnings.length}`, 'yellow');
    this.log(`❌ Errors: ${this.errors.length}`, 'red');

    if (this.errors.length === 0 && this.warnings.length === 0) {
      this.log('\n🎉 All validations passed!', 'green');
    } else if (this.errors.length === 0) {
      this.log('\n✅ Validation passed with warnings', 'yellow');
    } else {
      this.log('\n❌ Validation failed', 'red');
    }

    this.log('');
  }
}

// Run validator
if (require.main === module) {
  const validator = new Validator();
  validator.run();
}

module.exports = Validator;

File: scripts/validate.py
#!/usr/bin/env python3
"""
Validate repository structure and content.
Checks for:
- Missing files
- Broken links in README
- Inconsistent naming
- Empty files
- Missing metadata
"""

import os
import re
from pathlib import Path

class Colors:
    RED = '\033[91m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    BLUE = '\033[94m'
    RESET = '\033[0m'

def check_readme_links():
    """Check all links in README.md point to valid directories."""
    issues = []

    with open('README.md', 'r', encoding='utf-8') as f:
        content = f.read()

    # Find all markdown links to directories
    links = re.findall(r'\[.*?\]\((\.\/[^)]+)\)', content)

    for link in links:
        # Decode URL encoding
        path = link.replace('%20', ' ')
        path = path.replace('./', '')
        path = path.rstrip('/')

        if not os.path.exists(path) and path not in ['LICENSE.md', 'README.md']:
            issues.append(f"Broken link: {link} -> {path} does not exist")

    return issues

def check_directory_consistency():
    """Check all directories are listed in README."""
    issues = []

    # Get all directories
    dirs = [d for d in os.listdir('.') if os.path.isdir(d) and d not in ['.git', 'assets', '.github', 'scripts']]

    # Read README
    with open('README.md', 'r', encoding='utf-8') as f:
        readme_content = f.read()

    for directory in dirs:
        # Check if directory is mentioned in README (flexible matching)
        if directory not in readme_content and directory.replace(' ', '%20') not in readme_content:
            issues.append(f"Directory '{directory}' not listed in README.md")

    return issues

def check_empty_files():
    """Check for empty files."""
    issues = []

    for root, dirs, files in os.walk('.'):
        # Skip .git directory
        if '.git' in root:
            continue

        for file in files:
            if file.endswith(('.txt', '.json', '.md')):
                filepath = os.path.join(root, file)
                if os.path.getsize(filepath) == 0:
                    issues.append(f"Empty file: {filepath}")

    return issues

def check_file_naming_consistency():
    """Check for inconsistent file naming patterns."""
    warnings = []

    for root, dirs, files in os.walk('.'):
        if '.git' in root or 'scripts' in root:
            continue

        for file in files:
            # Check for files without extensions that should have them
            if not any(file.endswith(ext) for ext in ['.txt', '.json', '.md', '.png', '.jpg', '.svg']):
                if file not in ['.gitignore', 'LICENSE']:
                    warnings.append(f"File without standard extension: {os.path.join(root, file)}")

            # Check for inconsistent naming (mixing underscores and spaces)
            if '_' in file and ' ' in file:
                warnings.append(f"Inconsistent naming (mixed _ and space): {os.path.join(root, file)}")

    return warnings

def check_directory_structure():
    """Check each directory has appropriate files."""
    issues = []

    for item in os.listdir('.'):
        if os.path.isdir(item) and item not in ['.git', 'assets', '.github', 'scripts']:
            files = os.listdir(item)

            # Check if directory has any content files
            has_content = any(f.endswith(('.txt', '.json', '.md')) for f in files)
            if not has_content:
                issues.append(f"Directory '{item}' has no .txt, .json, or .md files")

    return issues

def main():
    """Run all validation checks."""
    print(f"{Colors.BLUE}{'='*60}{Colors.RESET}")
    print(f"{Colors.BLUE}Repository Validation{Colors.RESET}")
    print(f"{Colors.BLUE}{'='*60}{Colors.RESET}\n")

    all_issues = []
    all_warnings = []

    # Run checks
    checks = [
        ("README links", check_readme_links, True),
        ("Directory consistency", check_directory_consistency, True),
        ("Empty files", check_empty_files, True),
        ("File naming", check_file_naming_consistency, False),
        ("Directory structure", check_directory_structure, True)
    ]

    for check_name, check_func, is_error in checks:
        print(f"Checking {check_name}...", end=" ")
        results = check_func()

        if results:
            if is_error:
                all_issues.extend(results)
                print(f"{Colors.RED}✗ {len(results)} issue(s){Colors.RESET}")
            else:
                all_warnings.extend(results)
                print(f"{Colors.YELLOW}⚠ {len(results)} warning(s){Colors.RESET}")
        else:
            print(f"{Colors.GREEN}✓{Colors.RESET}")

    # Print detailed results
    if all_issues:
        print(f"\n{Colors.RED}Issues found:{Colors.RESET}")
        for issue in all_issues:
            print(f"  {Colors.RED}✗{Colors.RESET} {issue}")

    if all_warnings:
        print(f"\n{Colors.YELLOW}Warnings:{Colors.RESET}")
        for warning in all_warnings:
            print(f"  {Colors.YELLOW}⚠{Colors.RESET} {warning}")

    # Summary
    print(f"\n{Colors.BLUE}{'='*60}{Colors.RESET}")
    if not all_issues and not all_warnings:
        print(f"{Colors.GREEN}✓ All checks passed!{Colors.RESET}")
        return 0
    else:
        print(f"Total: {Colors.RED}{len(all_issues)} error(s){Colors.RESET}, {Colors.YELLOW}{len(all_warnings)} warning(s){Colors.RESET}")
        return 1 if all_issues else 0

if __name__ == "__main__":
    exit(main())
